chainerio/_context.py

```diff
@@ -1,116 +1,96 @@
+            return (path, "", True)
 
-        return ("posix", path, False)
-
-    def format_path(self, fs: IO, path: str) -> Tuple[str, bool]:
-        fs_type = fs.type
-        if fs_type in self.pattern_list.keys():
-            pattern = self.pattern_list[fs_type]
+        for fs_type, pattern in self.pattern_list.items():
             ret = pattern.match(path)
             if ret:
-                return (ret.groupdict()["path"], True)
-            else:
-                return (path, False)
-        else:
-            return (path, False)
+                return (fs_type, ret.groupdict()["path"], True)
 
-    def get_handler_from_path(self, path: str) -> Tuple[IO, str, bool]:
-        (fs_type, actual_path, is_URI) = self._determine_fs_type(path)
+        return ("posix", path, False)
 
-        handler = create_fs_handler(fs_type)
-        return (handler, actual_path, is_URI)
+    def get_or_create_cached_handler(self, fs_type: str) -> IO:
 
-    def get_handler_for_root(self,
-                             uri_or_handler_name: str) -> Tuple[IO, str, bool]:
-        if uri_or_handler_name in self.pattern_list.keys():
-            return (create_fs_handler(uri_or_handler_name), "", False)
+        self._handler_mt_lock.acquire()
+
+        if fs_type in self._handler_cache:
+            # get handler from cache
+            handler = self._handler_cache[fs_type]
         else:
-            (new_handler, actual_path, is_URI) = self.get_handler_from_path(
-                uri_or_handler_name)
-            new_handler.root = actual_path
-            return (new_handler, actual_path, is_URI)
+            # create a new handler
+            handler = create_fs_handler(fs_type)
+            self._handler_cache[fs_type] = handler
+
+        self._handler_mt_lock.release()
+
+        return handler
 
     def is_supported_scheme(self, scheme: str) -> bool:
         return scheme in self.scheme_list
 
 
 class DefaultContext(object):
     def __init__(self):
-        self._fs_handler_list = FileSystemDriverList()
-        self._root = ""
+        self._fs_handler_list = FileSystemHandlerList()
 
         self._default_context = \
-            self._fs_handler_list.get_handler_for_root("posix")[0]
+            self._fs_handler_list.get_or_create_cached_handler("posix")
 
     def set_root(self, uri_or_handler: Union[str, IO]) -> None:
-        # TODO(check) if root is directory
         if isinstance(uri_or_handler, IO):
             handler = uri_or_handler
-            self._root = ""
         else:
-            (handler, self._root, is_URI) = \
-                self.get_handler_by_name(uri_or_handler)
-        assert handler is not None
+            handler, path = self.get_or_create_default_handler(uri_or_handler)
+            handler.root = path
 
-        if self._root:
-            if not handler.isdir(self._root):
-                raise RuntimeError("the URI does not point to a directory")
+            if handler.root:
```

Comment:
This check must be done before setting `root` to prevent wrong path set to default handlers.

Reply:
resolved in 7c8df6c